src/hotspot/share/gc/shared/gcLogPrecious.cpp

```diff
@@ -77,11 +75,20 @@ void GCLogPrecious::vwrite_and_debug(LogTargetHandle log,
 }
 
 void GCLogPrecious::print_on_error(outputStream* st) {
-  if (_lines != NULL) {
-    MutexLocker locker(_lock, Mutex::_no_safepoint_check_flag);
-    if (_lines->size() > 0) {
+  if (_lines == NULL) {
+    return;
+  }
+
+  if (!_lock->trywait()) {
       st->print_cr("GC Precious Log:");
-      st->print_cr("%s", _lines->base());
-    }
+      st->print_cr(" ... skipping ...");
+      return;
   }
+
+  if (_lines->size() > 0) {
+    st->print_cr("GC Precious Log:");
+    st->print_cr("%s", _lines->base());
+  }
+
+  _lock->signal();
```

Comment:
As we discussed off-line, perhaps we might want to print something even if the log isn't initialized and/or is empty. Something like:
```
  st->print_cr("GC Precious Log:");

  if (_lines == NULL) {
    st->print_cr("<Not initialized>");
    return;
  }

  if (!_lock->trywait()) {
    st->print_cr("<Skipped>");
    return;
  }

  if (_lines->size() == 0) {
    st->print_cr("<Empty>");
  } else {
    st->print_cr("%s", _lines->base());
  }

  _lock->signal();
```
You decide. Looks good otherwise.

Reply:
Updated with suggestion. I also added extra newlines to make the output look pretty. That was needed because _lines->base() is always terminated with a newline.